### ISSUE TITLE
MLX SFT ignores training.max_grad_norm: sixteen transformers trainers clip, MLX does not, and nothing says so

### ISSUE THREAD

**Issue by Srinivasan8888** [MakazhanAlpamys/Soup#749]
## The defect

`max_grad_norm: float = Field(default=1.0, gt=0)` (`config/schema.py:1098`) is forwarded into `TrainingArguments` by sixteen transformers trainers, so the documented default clips every transformers run at 1.0. On `backend: mlx` it reaches nothing:

- no file under `src/soup_cli/trainer/mlx_*.py` mentions it;
- `mlx_lm 0.31.3`'s `TrainingArgs` has no such field, and its `trainer.py` contains no `clip` or `norm` call at all, so upstream does not clip either;
- `_check_unsupported()` (`mlx_sft.py:136`) does not list it, so unlike `training.seed` / `training.data_seed` nothing is printed.

The same config therefore trains **clipped on one backend and unclipped on the other**, with nothing in the output to tell them apart. This is the same family as #683 / #684 / #685 / #686, and it was not in any of their lists.

## Measured

On an M1, driving the real `mlx.optimizers` inside upstream's own compiled-step shape (`@partial(mx.compile, inputs=state, outputs=state)`, `trainer.py:246-248`), one step on a deliberately ill-conditioned batch:

```
                       unwrapped    clip=1.0    clip=0.01
SGD    |w - w0| =     388.025909    0.008490     0.000085
AdamW  |w - w0| =       0.126505    0.126504     0.126403
```

**The AdamW row is the honest bound on impact.** Adam normalises by its second moment, so clipping barely moves an AdamW step. What this actually rescues is `optimizer: sgd` / `lion`, and any run whose gradients spike — not every MLX run.

## Why it is fixable rather than only warnable

`mlx.optimizers.clip_grad_norm(grads, max_norm)` exists, and mlx-lm applies gradients through exactly one call — `optimizer.update(model, grad)` at `trainer.py:259`. So an optimizer that clips before delegating reaches it without forking upstream's loop. Everything but `update` has to delegate, because upstream reads `optimizer.state` (`trainer.py:246`) and `optimizer.learning_rate` (`trainer.py:337`) off the object it is handed.

The alternative — adding it to `_check_unsupported()` so the run says `MLX backend ignores: training.max_grad_norm` — is strictly worse when the fix is this small, but it is a legitimate option if you would rather not carry a wrapper.

## Acceptance criteria

1. A run on `backend: mlx` clips gradients at the configured `training.max_grad_norm`, asserted at the optimizer rather than in the plan or in metadata.
2. A different configured value produces a different clip, so the test cannot pass against a hardcoded default.
3. The underlying optimizer receives the **clipped** gradients, not the raw ones.
4. `optimizer.state` and `optimizer.learning_rate` remain reachable, since upstream reads both off the object it is given.
5. The norm that ran is recoverable from the output directory, because a Soup-side wrapper is otherwise invisible there.
6. Tests run on a CI runner with no MLX — no `importorskip` on the wiring assertions.

I have a branch that does this, with the measurements above and a five-mutation table. Happy to open it, or to switch to the warn-only option if you prefer that.

**Comment by MakazhanAlpamys** [comment]
Closed by #750, merged as `c7c6314a`. Credit to @Srinivasan8888.

MLX SFT now wraps its optimizer so `optim.clip_grad_norm` runs at the configured norm and the **clipped** gradients reach the inner optimizer. Attribute access delegates through, so the wrapper stays transparent to mlx-lm and to the schedule the optimizer carries.

**The framing in this issue was the useful part** and it is worth keeping: fifteen transformers trainers forward `max_grad_norm`, one site each, and MLX forwarded it nowhere — while `_check_unsupported` did not list it either, so the run did not even warn. A silently different optimization objective across backends is worse than an unsupported-feature error, because nothing tells the operator their MLX run is not the run their config describes.

**Six mutations at merge, six killed**, all on an interpreter where `find_spec("mlx") is None`, which is the condition every CI cell runs under. Two are worth naming because both leave a correct-looking `clip_grad_norm` call in place and change only what reaches the optimizer: applying the raw gradients and discarding the clip result, and clipping *after* the update rather than before. A test asserting "the clipper was called" would pass on either.

**What this does not prove, stated plainly:** no CI test asserts a clipped gradient magnitude. The tests use a recording fake and pin that the configured norm reaches the clipper and that the clipper's return value reaches the optimizer. The arithmetic evidence — SGD 388.03 to 0.0085 — is the contributor's own Apple Silicon measurement. What covers the real path instead is the `mlx-smoke` CI job, which drives the actual `soup train` CLI on macos-14 against real `mlx` and `mlx_lm` through `mx.compile`; it passed, so the wrapper survives tracing and upstream's handling.

This is the third setting in a week found validated, documented and read by nothing on the MLX path — after #683 and #686. #748 is about detecting that class systematically rather than one field at a time.
